pkg/limited/types.go

```diff
@@ -19,25 +28,31 @@ type Backend interface {
 	List(ctx context.Context, key, rangeEnd []byte, limit, revision int64) (int64, []*KeyValue, error)
 	Count(ctx context.Context, key, rangeEnd []byte, revision int64) (int64, int64, error)
 	Update(ctx context.Context, key, value []byte, revision, lease int64) (int64, bool, error)
-	Watch(ctx context.Context, key, rangeEnd []byte, revision int64) (<-chan []*Event, error)
+	WatcherGroup(ctx context.Context) (WatcherGroup, error)
 	DbSize(ctx context.Context) (int64, error)
-	CurrentRevision(ctx context.Context) (int64, error)
 	GetCompactRevision(ctx context.Context) (int64, int64, error)
 	DoCompact(ctx context.Context) error
 	Close() error
 }
 
-type KeyValue struct {
-	Key            string
-	CreateRevision int64
-	ModRevision    int64
-	Value          []byte
-	Lease          int64
+type WatcherGroup interface {
+	// Watch will add a watcher to the group. If startRevision is not 0, the first notification
+	// containing an update for this watcher will also contain all events from startRevision
+	// up to that notification.
+	Watch(watcherId int64, key, rangeEnd []byte, startRevision int64) error
+	Unwatch(watcherId int64)
+	Updates() <-chan WatcherGroupUpdate
```

Comment:
Another personal preference:
I think `WatchGroup` should come with an `Add` method (here the `Watch` is basically our `Add`), and it should add an instance of a `Watcher` interface. And the `Watcher` should implement a `Watch` method. So Maybe we can have something like:
```suggestion
type WatcherGroup interface {
	// Watch will add a watcher to the group. If startRevision is not 0, the first notification
	// containing an update for this watcher will also contain all events from startRevision
	// up to that notification.
	Add(Watcher) error
	Remove(Watcher)
	Updates() <-chan WatcherGroupUpdate
}

type Watcher interface {
	Watch(key, rangeEnd []byte, startRevision int64) error
	Id() int64
}
```

Reply:
I might not see it, but I think that this change is weird: it expects *the user* to pass a `Watcher` which will implement the `Watch` action, but how can the user implement a `Watcher` without relying on the internal details of the `WatcherGroup`?

To me the information flows in the other way: it is the `WatcherGroup` (in this case the one in `sqllog.go`) that knows that implementation details and know how to get those notifications. Am I missing something?

Reply:
Seems like I made some wrong assumptions about the responsibilities of a "watcher". I expected some sort of "Watcher" which "Watch"es something and is part of a "WatcherGroup". Ignore this comment then as I need to read the code more carefully.
From what I can see from the code, the watchers are controlled by the watcher group and don't act independently.
Initially, I thought of watchers as independent actors which can be configured according to the "group" they're a member of. If I understand this correctly, our watchers are more of a "receiver" rather than a "watcher". 
They're not actively looking for updates from a channel and they're not subscribed to a publisher or anything like that.
Information will get pushed to them in the `SQLLog` poll loop.
So basically the "watcher" here is the `SQLLOG`, not the individual instances of the `watcher` struct. Did I understand this correctly?

Reply:
Yeah basically the real flow is the `WatcherGroup` and a `Watcher` is one of the ranges the flow should notify about. I think that the misleading part is in the other code having another struct called `watcher`. I think there is a misnaming there.

Reply:
Got it. Thank you so much for clearing things for me @marco6! Does GH have a @marco++ or something similar? :D